### ISSUE TITLE
MCS+SMP: Missing remoteQueueUpdate somewhere?

### ISSUE THREAD

**Issue by Indanz** [seL4/seL4#941]
Setup:
- Passive server with bound notification.
- The notification has a scheduling context bound to CPU core 2. (This is the only task running on core 2.)
- All IPC callers are on core 1.

The passive server gets stuck when it is woken up by a signal while it was on core 1. I suspect when it just replied to a call.

A register dump shows the server and all callers are stuck in the system call, the server in `ReplyRecv`, the callers in `Call`.

Adding a `remoteQueueUpdate()` before the `break` in the `SysReplyRecv` case of `handleSyscall()` solves the problem for me (with fastpath disabled for easier debugging).

I'll continue debugging tomorrow and update this issue when more is known.

Anyone any idea what might be up?

**Comment by kent-mcleod** [comment]
> Anyone any idea what might be up?

In `receiveSignal` when `NtfnState_Active`, `maybeDonateSchedContext()` will call migrateTCB, but the TCB will never be inserted into any scheduler queues on the other core. So there's missing calls to SCHED_APPEND for triggering the schedule on the other core, and also to rescheduleRequired() for rescheduling on the current core. It doesn't appear that this scenario is being handled by the kernel at the moment.  We need to add a test case for this: when ksCurThread is passive, receiving a endpoint or notification with a sc for a different core will migrate and reschedule the thread on the other core.

**Comment by Indanz** [comment]
Thanks for looking into this.

`maybeDonateSchedContext()` calls:
- `schedContext_donate`, which assigns the SC to the TCB (and does migrateTCB).
- `schedContext_resume`, which may postpone the SC if it can't run now.

It may make more sense to do the migrateTCB in the resume instead of the donate, as well as possibly any new check.

Though I don't think my case comes via `receiveSignal`, but via `receiveIPC`, which calls `completeSignal` if a signal is pending, but from there on the code is almost the same (there is a subtle difference between the sporadic check of `NtfnState_Active` compared to the one in `completeSignal`).

**Comment by Indanz** [comment]
The code in `schedContext_bindTCB` is relevant too. It bugs me that all callers of `schedContext_resume` handle this case slightly differently (like calling enqueue and rescheduleRequired in different orders).

Should we add the following check somewhere or is something else needed?
```
if (isSchedulable(target)) {
    possibleSwitchTo(target);
}
```

**Comment by Indanz** [comment]
Perhaps:

```
diff --git a/src/object/notification.c b/src/object/notification.c
index de96d7f4caf9..f8c2db447ff3 100644
--- a/src/object/notification.c
+++ b/src/object/notification.c
@@ -235,7 +235,9 @@ void receiveSignal(tcb_t *thread, cap_t cap, bool_t isBlocking)
             notification_ptr_get_ntfnMsgIdentifier(ntfnPtr));
         notification_ptr_set_state(ntfnPtr, NtfnState_Idle);
 #ifdef CONFIG_KERNEL_MCS
-        maybeDonateSchedContext(thread, ntfnPtr);
+        MCS_DO_IF_SC(thread, ntfnPtr, {
+                possibleSwitchTo(thread);
+        })
         // If the SC has been donated to the current thread (in a reply_recv, send_recv scenario) then
         // we may need to perform refill_unblock_check if the SC is becoming activated.
         if (thread->tcbSchedContext != NODE_STATE(ksCurSC) && sc_sporadic(thread->tcbSchedContext)) {
@@ -309,7 +311,9 @@ void completeSignal(notification_t *ntfnPtr, tcb_t *tcb)
         setRegister(tcb, badgeRegister, badge);
         notification_ptr_set_state(ntfnPtr, NtfnState_Idle);
 #ifdef CONFIG_KERNEL_MCS
-        maybeDonateSchedContext(tcb, ntfnPtr);
+        MCS_DO_IF_SC(tcb, ntfnPtr, {
+                possibleSwitchTo(tcb);
+        })
         if (sc_sporadic(tcb->tcbSchedContext)) {
             sched_context_t *sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
             if (tcb->tcbSchedContext == sc && tcb->tcbSchedContext != NODE_STATE(ksCurSC)) {
```

**Comment by Indanz** [comment]
I can confirm that the above patch is not sufficient, so either something else is missing, or this is the wrong place. It does fix the original problem I had, but tasks still hang under load with many cross-core passive calls.

Those hangs disappear if I add an `remoteQueueUpdate()` in to the send signal path where thread sate is `!= ThreadState_BlockedOnReceive` (using the signal fastpath for easier testing here), but I fear that's just papering over the real bug.

**Comment by kent-mcleod** [comment]
`SCHED_ENQUEUE_CURRENT_TCB` could probably have an assert that the current TCB still has the same affinity of the current core. I agree with @Indanz's comments in #986 and the examples here that by the time that `schedule()` is called, the current TCB should have been correctly inserted in the queue on the remote core and the remoteIPI flags set up.  `schedule()` should then not cause the current TCB to be reinserted back onto the current core's queue.  
There's probably no existing pattern for migrating the current thread to another core?
What ways can ksCurThread be synchronously migrated to a different core?:
- SchedControl_Configure can update the current thread's SC to a different core,
- SchedContext donation via endpoints and notifications can move a TCB to a different core.
-

**Comment by Indanz** [comment]
I agree that an assert in `SCHED_ENQUEUE_CURRENT_TCB` would be a good idea.

The original issue is about activating a task cross-core, not about migrating `ksCurThread`, although that would be one way to activate a thread on another core. We should make a list of ways to activate a thread on another core in such way that it should be picked for scheduling. Whether the TCB needs to be migrated or not doesn't matter for this list, it's just that there is a higher chance that that case is handled incorrectly.

In below scenarios all activated tasks are assumed to have the highest priority on the remote core:

- SchedContext_Bind on a TCB can give a task budget to run on another core.
- SchedContext_Bind on a signalled NF bound to a TCB can give a task budget to run on another core.
- SchedControl_Configure can migrate a task's SC to a different core (either the current task or another one).
- SchedControl_Configure can update a remote task's SC to have budget to run immediately.
- seL4_TCB_SetPriority can increase a remote task's priority.
- A suspended remote task can be resumed.
- seL4_SchedContext_YieldTo can give budget to a remote task.
- A blocking remote task can be activated by a signal.
- A blocking remote task can be activated by a call.
- A blocking remote task can be activated by a reply.
- A blocking passive remote task can be activated by a signal, triggering core migration for the TCB.
- A blocking passive remote task can be activated by a call, triggering core migration for the TCB.

And the same, but by making the currently active remote task not run any more and activating the next ready task:

- seL4_SchedContext_Unbind and seL4_SchedContext_UnbindObject can take away a remote task's SC.
- SchedControl_Configure can migrate a remote task.
- SchedControl_Configure can take away a remote task's budget.
- seL4_TCB_SetPriority can decrease a remote task's priority.
- seL4_TCB_Suspend can suspend a remote task.
- Remote task's TCB or SC objects can be destroyed.

I think there should be tests for each of these cases in sel4test. For completeness, all cases where a remote task is migrated to the current core could also be added. This find some SMP bugs in regard to scheduling and would give some reassurance whether the fix for this issue is complete or not.

**Comment by kent-mcleod** [comment]
> I think there should be tests for each of these cases in sel4test. For completeness, all cases where a remote task is migrated to the current core could also be added. This find some SMP bugs in regard to scheduling and would give some reassurance whether the fix for this issue is complete or not.

This makes sense.  If we started a branch in sel4test would anyone be interested in helping add testcases to work through the enumerated list?
